app/views/course/assessment/answer/programming/_test_cases.html.slim

```diff
@@ -1,127 +1,50 @@
-- test_cases = question.test_cases
-- public_test_cases = test_cases.select(&:public_test?)
-- private_test_cases = test_cases.select(&:private_test?)
-- evaluation_test_cases = test_cases.select(&:evaluation_test?)
-- submission = answer.submission
 - auto_grading = answer.auto_grading ? answer.auto_grading.actable : nil
 - answer_test_results_hash = auto_grading ? \
     auto_grading.test_results.map { |result| [result.test_case, result] }.to_h : \
     {}
+- is_grader = can?(:grade, answer.answer)
+- test_cases_by_type = question.test_cases_by_type
 
+/ Test case results table
 h3 = t('.test_cases')
 
-table.table.table-striped.table-hover
-  thead
-    tr
-      - if can?(:grade, answer.answer)
-        th = Course::Assessment::Question::ProgrammingTestCase.human_attribute_name(:identifier)
-      th = Course::Assessment::Question::ProgrammingTestCase.human_attribute_name(:expression)
-      th = Course::Assessment::Question::ProgrammingTestCase.human_attribute_name(:expected)
-      - if can?(:grade, answer.answer)
-        th = t('.output')
-      - if @submission.attempting?
-        th = Course::Assessment::Question::ProgrammingTestCase.human_attribute_name(:hint)
-      th = t('.passed')
-  tbody
-    tr
-      - if can?(:grade, answer.answer)
-        th colspan=5
-          = t('.public')
-      - else
-        th colspan = 4
-          = t('.public')
-    - public_test_cases.each do |test_case|
-      - test_case_result = answer_test_results_hash[test_case]
-      = content_tag_for(:tr, test_case) do
-        - if can?(:grade, answer.answer)
-          th = format_html(test_case.identifier)
-        td = format_html(test_case.expression)
-        td
-          div.expected = format_html(test_case.expected)
-        - if can?(:grade, answer.answer)
-          td = simple_format(get_output(test_case_result))
-        - if @submission.attempting?
-          td
-            - if !test_case_result || !test_case_result.passed?
-              = simple_format(get_hint(test_case, test_case_result))
-        td
-          - if test_case_result && test_case_result.passed?
-            = fa_icon 'check'.freeze
-          - elsif test_case_result
-            = fa_icon 'times'.freeze
+= render partial: 'course/assessment/answer/programming/test_cases_of_type',
```

Comment:
I think we can shift the check of `if test_cases` out here. 

Assuming you constructed the hash above - so something like this:
```ruby
h3 = t('.test_cases')

- if test_cases_and_answers.keys.include?('public_test')
  = render partial: 'test_cases_of_type', locals: { test_cases_and_answers: test_cases_and_answers['public_test'], type_heading: t('.public'), is_grader: is_grader }
```